Multitenancy for managed clusters

shysank opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md). -->

 <!-- Please label this pull request according to what type of issue you are addressing (see ../CONTRIBUTING.md) -->
**What type of PR is this?**
/kind feature
<!--
Add one of the following kinds:
/kind feature
/kind bug
/kind api-change
/kind cleanup
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind flake
/kind other
-->

**What this PR does / why we need it**:
This PR implements multitenancy for azure managed clusters largely building upon the work done in #977

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1153 

**Special notes for your reviewer**:
A couple of things to note:

- `NetworkInfrastructureReadyCondition` not implemented. There are no conditions right now for managed clusters. So, will create a followup issue for this
-  There is a separate `AzureIdentityReconciler` for reconciling `AzureIdentity` and `AzureIdentityBinding` in exp/controllers. which does exactly the same thing as the one already there in controllers with one difference being it only deletes identities created by managed clusters. This should eventually merged into a single controller once managed clusters are promoted from experimental.

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Multitenancy for managed clusters.
```

shysank commented:
/retest

nader-ziada commented on azure/scope/identity.go:
```diff
@@ -37,17 +38,36 @@ import (
 )
 
-// AzureCredentialsProvider provides
+// CredentialsProvider ...
```
can you add a description

nader-ziada commented on exp/api/v1alpha4/azuremanagedcontrolplane_types.go:
```diff
@@ -82,6 +82,10 @@ type AzureManagedControlPlaneSpec struct {
 	// +optional
 	LoadBalancerSKU *string `json:"loadBalancerSKU,omitempty"`
+
+	// IdentityRef is a reference to a AzureIdentity to be used when reconciling this cluster
```
```suggestion
	// IdentityRef is a reference to a AzureClusterIdentity to be used when reconciling this cluster
```

nader-ziada commented on exp/controllers/azuremangedcontrolplane_controller.go:
```diff
@@ -161,8 +163,29 @@ func (r *AzureManagedControlPlaneReconciler) Reconcile(ctx context.Context, req
 
 	log = log.WithValues("machinePool", ownerPool.Name)
 
+	// check if the control plane's namespace is allowed for this identity and update owner references for the identity.
```
why are you updating the Owner reference of the identity? what if this identity is used by two different clusters?

shysank replied:
We update the ownerReferences for `AzureClusterIdentity` which has 1:1 relationship with the cluster. We do similarly in [azure cluster controller](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/83853b7148355637cd8329e9edf64b67ffa9af79/controllers/azurecluster_controller.go#L211).

nader-ziada replied:
yeah, you are right, my bad!

nader-ziada reviewed (commented):
is it possible to have an e2e test for this?

CecileRobertMichon commented on exp/controllers/azureidentity_controller.go:
```diff
@@ -0,0 +1,209 @@
+)
+
+// AzureIdentityReconciler reconciles azure identity objects
+type AzureIdentityReconciler struct {
```
how do we distinguish this from https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/controllers/azureidentity_controller.go once we move managedclusters out of exp/ ?

nader-ziada replied:
both should get merged in one at that point, we had a chat about and I felt a controller in the controllers folder should not check types in the `exp` folder

nader-ziada commented:
I think this is good to go
@devigned do you have any thoughts?

devigned reviewed (approved):
lgtm

The issue with the repeated identity controller should either be refactored here or in #1389.

CecileRobertMichon commented:
> The issue with the repeated identity controller should either be refactored here or in #1389

I don't intend to refactor exp/ controllers or dir structure in #1389, IMO we should tackle it here. The identity feature is not experimental so it shouldn't leave in exp/.

nader-ziada commented:
> > > The issue with the repeated identity controller should either be refactored here or in #1389
> > 
> > 
> > I don't intend to refactor exp/ controllers or dir structure in #1389, IMO we should tackle it here. The identity feature is not experimental so it shouldn't leave in exp/.
> 
> Two solutions I can think of:
> 
> 1. Reconcile identities created by amcp along with azure cluster. This is what Nader and I discussed; we didn't want to import exp api types inside the main controllers package.
> 2. Ignore identities created by amcp, and just make sure they don't get deleted. A hacky solution until we remove it from exp, lots of dangling identities will be there. There may also be issues when we create a new amcp with the same in future.
> 
> Both of these are not ideal solutions. The main problem I see is, although the identity feature is not experimental, in this case it's tightly coupled with the exp api types which makes it difficult to move it out of exp. wdyt?

The identity controller in the `exp` folder is just to deal with identities related to managedcontrolplane clusters, I don't understand why it is an issue to keep it there. If the issue is about duplication of code we can try to refactor some of the code in a util package and use it in both, but I still think they should be two controllers

CecileRobertMichon commented:
> Reconcile identities created by amcp along with azure cluster. This is what Nader and I discussed; we didn't want to import exp api types inside the main controllers package.

Why is importing exp types in the main controller a bad thing? As long as it's all behind feature gates that should be fine and, IMO, cleaner than duplicating code that might diverge in the future. We already do this in CAPI, eg: 
https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/cluster_controller.go#L365-L373

nader-ziada commented:
> > Reconcile identities created by amcp along with azure cluster. This is what Nader and I discussed; we didn't want to import exp api types inside the main controllers package.
> 
> Why is importing exp types in the main controller a bad thing? As long as it's all behind feature gates that should be fine and, IMO, cleaner than duplicating code that might diverge in the future. We already do this in CAPI, eg:
> https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/cluster_controller.go#L365-L373

I didn't think this was allowed, in that case, my bad for asking Shyam to not do that and have a 2nd controller, we can just combine them then

shysank commented:
Thanks @CecileRobertMichon and @nader-ziada for the feedback. I'll update the existing identity_controller to handle aks identities, and remove the new one.

nader-ziada commented on controllers/azureidentity_controller.go:
```diff
@@ -124,18 +154,49 @@ func (r *AzureIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+
+		// Only delete bindings that are created by AzureCluster
+		if cluster.Spec.InfrastructureRef.Kind != "AzureCluster" {
+			continue
```
do we still need this?

shysank replied:
Good catch, I've removed it. Thanks!

shysank commented:
/test pull-cluster-api-provider-azure-e2e-windows

shysank commented:
/test pull-cluster-api-provider-azure-e2e

nader-ziada commented:
thank you @shysank 

/lgtm

nader-ziada commented:
/lgtm cancel

forgot @shysank is still adding an e2e test, which is great by the way :)

CecileRobertMichon commented:
@shysank if you're adding an e2e test, might actually solve #611 at the same time (given #1360 I don't think we need a separate e2e test for multi-tenancy managed clusters vs. managed clusters using env vars for creds)

shysank commented:
@CecileRobertMichon @nader-ziada , I've added an e2e for aks clusters with multi tenancy. PTAL, whenever you get a chance. One thing to note is that AKS doesn't support kubernetes `1.19.7`, so I have updated it to `1.19.9`

CecileRobertMichon commented:
> @CecileRobertMichon @nader-ziada , I've added an e2e for aks clusters with multi tenancy. PTAL, whenever you get a chance. One thing to note is that AKS doesn't support kubernetes 1.19.7, so I have updated it to 1.19.9

AKS has its own k8s support matrix so I recommend not reusing the same var, otherwise we might get into situations where we bump the k8s version but AKS doesn't support it yet, amd vice versa. 

Furthermore, it might make sense to look into an alternative to a statically defined k8s version here and instead define a minor version but get the latest supported patch at runtime (you can do `az aks get-versions -l location`, and there's probably an SDK equivalent). Otherwise, this will quickly become a pain to maintain as we'll have to constantly play catch up with AKS patch version support (they only support 2 patches of each minor at any given time https://docs.microsoft.com/en-us/azure/aks/supported-kubernetes-versions), which will regularly cause the test to break. Thoughts?

shysank commented:
> Furthermore, it might make sense to look into an alternative to a statically defined k8s version here and instead define a minor version but get the latest supported patch at runtime (you can do `az aks get-versions -l location`, and there's probably an SDK equivalent). Otherwise, this will quickly become a pain to maintain as we'll have to constantly play catch up with AKS patch version support (they only support 2 patches of each minor at any given time https://docs.microsoft.com/en-us/azure/aks/supported-kubernetes-versions), which will regularly cause the test to break. Thoughts?

yeah, makes sense to make it not static. How about, get `major.minor` version from `azure-dev.yaml`, and pick the latest patch supported by `az sk get-vesions -l location` for that `major.minor`? In our case, it'll be the latest patch for `1.19` supported by `az aks get-versions -l location`

CecileRobertMichon commented:
> yeah, makes sense to make it not static. How about, get major.minor version from azure-dev.yaml, and pick the latest patch supported by az sk get-vesions -l location for that major.minor? In our case, it'll be the latest patch for 1.19 supported by az aks get-versions -l location

SGTM, but let's add a separate K8s version var for AKS in azure-dev.yaml, in case we want to start testing new minor versions and AKS doesn't support it yet.

nader-ziada commented:
> e K8s version var for AKS in azure-dev.yaml, in case we want to start testing new minor versions and AKS doesn't support it yet.

+1 to adding a var,

also the tiltfile has this var for aks ` "aks_kubernetes_version": "v1.20.5"`

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/Tiltfile#L22

nader-ziada commented on test/e2e/azure_test.go:
```diff
@@ -454,6 +454,55 @@ var _ = Describe("Workload cluster creation", func() {
+			}, result)
+		})
+	})
+
```
test looks good

any validation we can do after its created?

shysank replied:
Do you mean something like this https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/4312c6c386e381f11708a8fceb395edbf9f2b89c/test/e2e/azure_identity.go#L44? we could something similar for agent pools.

nader-ziada replied:
yeah some like that, or whatever way you think you can check and validated things got created as expected

shysank commented:
/pull-cluster-api-provider-azure-e2e-windows

nader-ziada commented:
nice work @shysank 

lgtm 

will let other folks also take a look

CecileRobertMichon commented on test/e2e/config/azure-dev.yaml:
```diff
@@ -86,9 +86,15 @@ providers:
       targetName: "cluster-template-external-cloud-provider.yaml"
+    - sourcePath: "${PWD}/templates/test/ci/cluster-template-prow-aks-multi-tenancy.yaml"
+      targetName: "cluster-template-aks-multi-tenancy.yaml"
+    replacements:
```
what's this? rebase errata from #1332 ?

nader-ziada replied:
I noticed it in my PR (cluster identity) as well after I rebased, didn't realize it should be removed

shysank replied:
I actually thought this was needed by the e2e framework. I'll remove it if not needed.

nader-ziada replied:
I think it came from the rebase, I had the same happen, its not in the file in the `main` branch

CecileRobertMichon replied:
yeah the replacements were moved above inside the manager in another PR (which is where they should have been since the beginning)

shysank replied:
Oh my bad, I was thinking you were talking about 
```
    - sourcePath: "${PWD}/templates/test/ci/cluster-template-prow-aks-multi-tenancy.yaml"      
        targetName: "cluster-template-aks-multi-tenancy.yaml"
```

Let me remove the replacements block.

CecileRobertMichon reviewed (commented):
/cc @alexeldeib

CecileRobertMichon commented:
/cc @LochanRn

alexeldeib commented on controllers/azureidentity_controller.go:
```diff
@@ -96,20 +110,36 @@ func (r *AzureIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+			identityOwner = azureManagedControlPlane
+			err = r.Get(ctx, req.NamespacedName, azureManagedControlPlane)
+			if err != nil && apierrors.IsNotFound(err) {
+				r.Recorder.Eventf(azureCluster, corev1.EventTypeNormal, "AzureClusterObjectNotFound", err.Error())
```
delete L128? we're only looking up AMCP, AzureCluster shouldn't be hitting not found here.

alexeldeib replied:
or else need a separate var to avoid shadowing the error from L120

alexeldeib commented on controllers/azureidentity_controller.go:
```diff
@@ -96,20 +110,36 @@ func (r *AzureIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+			identityOwner = azureManagedControlPlane
+			err = r.Get(ctx, req.NamespacedName, azureManagedControlPlane)
+			if err != nil && apierrors.IsNotFound(err) {
+				r.Recorder.Eventf(azureCluster, corev1.EventTypeNormal, "AzureClusterObjectNotFound", err.Error())
```
(deleted and then decided this was valid, sorry for spam)

err is shadowed, but we know it's not found in both cases. let's either not shadow it and log correct errors, or not log the full error in both cases.

alexeldeib commented on go.mod:
```diff
@@ -31,6 +31,7 @@ require (
 	go.opentelemetry.io/otel/exporters/trace/jaeger v0.13.0
 	go.opentelemetry.io/otel/sdk v0.13.0
 	golang.org/x/crypto v0.0.0-20210220033148-5ea612d1eb83
+	golang.org/x/mod v0.4.2
```
I suspect we don't want this dep in go.{mod,sum}? https://pkg.go.dev/golang.org/x/mod not sure where that got pulled in?

shysank replied:
This is used for [semver](https://github.com/shysank/cluster-api-provider-azure/blob/a886e433bb2543d2144b79eb6fd6a6d1ade5403b/test/e2e/aks.go#L30) parsing for calculating [aks k8s version](https://github.com/shysank/cluster-api-provider-azure/blob/a886e433bb2543d2144b79eb6fd6a6d1ade5403b/test/e2e/aks.go#L225).

alexeldeib commented on test/e2e/aks.go:
```diff
@@ -0,0 +1,276 @@
+limitations under the License.
+*/
+
+package e2e
```
this whole file 😍

alexeldeib commented on test/e2e/aks.go:
```diff
@@ -0,0 +1,276 @@
+	containerServiceClient := containerservice.NewContainerServicesClient(subscriptionID)
+	containerServiceClient.Authorizer = authorizer
+
+	result, err := containerServiceClient.ListOrchestrators(ctx, location, ManagedClustersResourceType)
```
nice

alexeldeib commented:
very minor comments, code looks good so far. 

thank you *so* much for doing the work to add e2e! love seeing that in the output: 
![image](https://user-images.githubusercontent.com/6800857/121966665-e1c0f200-cd23-11eb-844c-2686513b0ae5.png)

alexeldeib commented on controllers/azureidentity_controller.go:
```diff
@@ -96,20 +111,38 @@ func (r *AzureIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Reques
-	if err != nil {
-		if apierrors.IsNotFound(err) {
+	if err != nil && apierrors.IsNotFound(err) {
+		if feature.Gates.Enabled(feature.AKS) {
```
instead of doing this check, we could probably use an unstructured client to lookup whatever GVK/object we get, pull `spec.identity` and drop the feature gate check + error handling.

fine to leave for follow up imo

alexeldeib commented:
this looks fairly good to go? anything you're still working on?

shysank commented:
> this looks fairly good to go? anything you're still working on?
Nope, I think we can this one in and create a follow up issue for more e2e scenarios.

alexeldeib commented:
/lgtm

nice work!

CecileRobertMichon reviewed (commented):
/lgtm
/assign @nader-ziada

nader-ziada commented:
👏👏

/approve

nader-ziada commented:
@shysank you are going to have to rebase so we can get this in

shysank commented:
/test pull-cluster-api-provider-azure-e2e

nader-ziada commented:
/test pull-cluster-api-provider-azure-e2e

nader-ziada commented:
/lgtm
